sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/PortEvaluationStage.java

```diff
@@ -0,0 +1,125 @@
+            OfferRequirement offerRequirement,
+            OfferRecommendationSlate offerRecommendationSlate) throws OfferEvaluationException {
+        // If this is from an existing pod with the dynamic port already assigned and reserved, just keep it.
+        Protos.CommandInfo commandInfo = getTaskName().isPresent() ?
+                offerRequirement.getTaskRequirement(getTaskName().get()).getTaskInfo().getCommand() :
+                offerRequirement.getExecutorRequirementOptional().get().getExecutorInfo().getCommand();
+        String taskPort = CommandUtils.getEnvVar(commandInfo, portName);
+        Integer assignedPort = port;
+        if (assignedPort == 0 && taskPort != null) {
+            assignedPort = Integer.parseInt(taskPort);
+        }
+
+        if (assignedPort == 0) {
+            // We don't want to dynamically consume a port that's explicitly claimed by this pod accidentally, so
+            // compile a list of those to check against the offered ports.
+            Set<Integer> consumedPorts = new HashSet<>();
+            for (Protos.Resource resource : offerRequirement.getResources()) {
+                if (resource.getName().equals(Constants.PORTS_TYPE)) {
+                    resource.getRanges().getRangeList().stream()
+                            .flatMap(r -> IntStream.rangeClosed((int) r.getBegin(), (int) r.getEnd()).boxed())
+                            .forEach(consumedPorts::add);
+                }
+            }
+
+            Protos.Value availablePorts = offerResourcePool.getUnreservedMergedPool().get(Constants.PORTS_TYPE);
+            Optional<Integer> dynamicPort = Optional.empty();
+            if (availablePorts != null) {
+                dynamicPort = availablePorts.getRanges().getRangeList().stream()
+                        .flatMap(r -> IntStream.rangeClosed((int) r.getBegin(), (int) r.getEnd()).boxed())
+                        .filter(p -> !consumedPorts.contains(p))
+                        .findFirst();
+            }
+
+            if (!dynamicPort.isPresent()) {
+                throw new OfferEvaluationException(String.format(
+                        "No ports were available for dynamic claim in offer: %s",
+                        offerResourcePool.getOffer().toString()));
+            }
+            assignedPort = dynamicPort.get();
+        }
+
+        super.setResourceRequirement(getDynamicPortRequirement(getResourceRequirement(), assignedPort));
+        super.evaluate(offerResourcePool, offerRequirement, offerRecommendationSlate);
+    }
+
+    @Override
+    protected void setProtos(OfferRequirement offerRequirement, Protos.Resource resource) {
+        long port = resource.getRanges().getRange(0).getBegin();
+
+        if (getTaskName().isPresent()) {
+            String taskName = getTaskName().get();
+            Protos.TaskInfo taskInfo = offerRequirement.getTaskRequirement(taskName).getTaskInfo();
+            Protos.TaskInfo.Builder taskInfoBuilder = taskInfo.toBuilder();
```

Comment:
It looks like we don't need to copy the `TaskInfo` here, as the `ports` being manipulated below is itself a copy coming from the `TaskInfo`. Could move this `toBuilder()` call immediately above `setResource()` below at L90.